test/jdk/java/lang/instrument/RetransformRecordAnnotation.java

```diff
@@ -0,0 +1,174 @@
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @bug 8315575
+ * @summary test that records with invisible annotation can be retransformed
+ *
+ * @library /test/lib
+ * @run shell MakeJAR.sh retransformAgent
+ * @run main/othervm -javaagent:retransformAgent.jar -Xlog:redefine+class=trace RetransformRecordAnnotation
+ */
+
+import java.io.File;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+import java.lang.instrument.ClassFileTransformer;
+import java.nio.file.Files;
+import java.security.ProtectionDomain;
+
+public class RetransformRecordAnnotation extends AInstrumentationTestCase {
+
+    @Target({ElementType.TYPE_PARAMETER, ElementType.TYPE_USE})
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface RuntimeTypeAnno {}
+
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface RuntimeParamAnno {
+        String s() default "foo";
+    }
+
+    @Target({ElementType.TYPE_PARAMETER, ElementType.TYPE_USE})
+    @Retention(RetentionPolicy.CLASS)
+    @interface ClassTypeAnno {}
+
+    @Retention(RetentionPolicy.CLASS)
+    @interface ClassParamAnno {
+        String s() default "bar";
+    }
+
+    @RuntimeTypeAnno
+    @RuntimeParamAnno(s = "1")
+    public record VisibleAnnos(@RuntimeTypeAnno @RuntimeParamAnno(s = "2") Object o, Object other) {
+    }
+
+    @ClassTypeAnno
+    @ClassParamAnno(s = "3")
+    public record InvisibleAnnos(@ClassTypeAnno @ClassParamAnno(s = "4") Object o, Object other) {
+    }
+
+    @RuntimeTypeAnno
+    @RuntimeParamAnno(s = "5")
+    @ClassTypeAnno
+    @ClassParamAnno(s = "6")
+    public record MixedAnnos(@RuntimeTypeAnno @RuntimeParamAnno(s = "7")
+                             @ClassTypeAnno @ClassParamAnno(s = "8") Object o, Object other) {
+    }
+
+    public static void main (String[] args) throws Throwable {
+        ATestCaseScaffold test = new RetransformRecordAnnotation();
+        test.beVerbose();
+        test.runTest();
+    }
+
+    // for Transformer
+    private String targetClassName;
+    private byte[] seenClassBytes;
+    private byte[] newClassBytes;
+
+    public RetransformRecordAnnotation() throws Throwable {
+        super("RetransformRecordAnnotation");
+    }
+
+    private void log(Object o) {
+        System.out.println(String.valueOf(o));
+    }
+
+    // Retransforms target class using provided class bytes;
+    // Returns class bytes passed to the transformer.
+    private byte[] retransform(Class targetClass, byte[] classBytes) throws Throwable {
+        targetClassName = targetClass.getName();
+        seenClassBytes = null;
+        newClassBytes = classBytes;
+        fInst.retransformClasses(targetClass);
+        assertTrue(targetClass.getName() + " was not seen by transform()", seenClassBytes != null);
```

Comment:
Nit: I guess, the `targetClassName` was intended to be used in place of `targetClass.getName()`.

Reply:
The `retransform()` method returns an array of bytes. There is also a comment about it at the line 101. But the result has never been checked (please, see lines: 116, 123, 129, 135). This creates some confusion and questions why the return value is needed.
The parameter `null` at lines 123, 129, 135 is also confusing and need some explanation.
As I read the code, the method  `retransform()` is saving the parameter `classBytes` in the `newClassBytes` field. The field `newClassBytes` value is then returned by the `transform()` method.
But the `ClassFileTransformer.html#transform` states this:
`Returns: a well-formed class file buffer (the result of the transform), or null if no transform is performed`.

So that, when the `null` is returned by the `transform()` then it means there was no real transform.
Could you comment on this? Do I miss anything here?

Please, see: [ClassFileTransformer.html#transform](https://docs.oracle.com/en/java/javase/21/docs/api/java.instrument/java/lang/instrument/ClassFileTransformer.html#transform(java.lang.Module,java.lang.ClassLoader,java.lang.String,java.lang.Class,java.security.ProtectionDomain,byte%5B%5D))

Also, the comment at line 122 is not needed at this form:
```
            // Ensure retransformation does not fail with ClassFormatError.
```
Sorry, I was not clear when I've asked to add a comment at this point. I wanted a clarification about what does passing the `null` mean in terms of transformation.
My understanding is that there is no real transformation when the `null` is returned but the `Instrumentation` mechanism is still working and can fail and we want to test/verify it.  So, passing a Null-transformer is enough for our testing. Is it correct?
Please, keep in mind that few people have some knowledge about the re-transformation details and would appreciate any clarification here.

Reply:
Instrumentation.retransformClasses logic can be described by the following pseudo-code:
```
byte[] newClassBytes = JvmtiClassFileReconstituter.reconstitute(klass);
for (Transformer tr = firstTransformer(); tr != null; tr = tr->next()) {
  byte[] transformerClassBytes = tr.transform(newClassBytes);
  if (newClassBytes != null) {
    newClassBytes = transformerClassBytes;
  }
}
```
Then newClassBytes is parsed, verified and is used to redefine the class.

We need to verify that JvmtiClassFileReconstituter produces correct class bytes, so test transformer can return null or passed classfileBuffer.
Actually the issue can be reproduced without ClassFileTransformer at all (and this is mentioned in the CR) - in the case reconstituted class bytes is considered as a result of the instrumentation.

I've updated the test, hope it's clearer now.
`targetClassName`/`seenClassBytes`/`newClassBytes` was used by Transformer, I moved them to the class.
Please let me know if you think some additional comments would be useful.

Reply:
Thank you for the update, it is a nice change.
I'd also suggest to add a comment before line 115, something like this:
```
  // Below the null is passed as the classBytes argument which means there won't be any transformation.
  // However, it is enough for testing purposes as the JvmtiClassFileReconstituter still involved
  // into preparation of the initial classfile bytecodes.
```

Reply:
Thank you for the update.